src/providers/azure.rs

```diff
@@ -0,0 +1,519 @@
+//! let my_machine = vms.get("my machine").unwrap();
+//! let (stdout, stderr) = my_machine.ssh.as_ref().unwrap().cmd("echo \"Hello, Azure\"").unwrap();
+//! println!("{}", stdout);
+//! ```
+
+use crate::ssh;
+use educe::Educe;
+use failure::{bail, Error};
+use std::collections::HashMap;
+use std::sync::Arc;
+
+/// A descriptor for a single Azure VM type.
+///
+/// The default is an UbuntuLTS, Standard_DS1_V2 VM in the East US region.
+#[derive(Clone, Educe)]
+#[educe(Debug)]
+pub struct Setup {
+    region: Region,
+    instance_type: String,
+    image: String,
+    username: String,
+    #[educe(Debug(ignore))]
+    setup_fn:
+        Option<Arc<dyn Fn(&mut ssh::Session, &slog::Logger) -> Result<(), Error> + Send + Sync>>,
+}
+
+impl Default for Setup {
+    fn default() -> Self {
+        Setup {
+            region: "eastus".parse().unwrap(),
+            instance_type: "Standard_DS1_v2".to_string(),
+            image: "UbuntuLTS".to_string(),
+            username: "ubuntu".to_string(),
+            setup_fn: None,
+        }
+    }
+}
+
+impl super::MachineSetup for Setup {
+    type Region = Region;
+
+    fn region(&self) -> Self::Region {
+        self.region
+    }
+}
+
+impl Setup {
+    /// See also [`Region`].
+    pub fn region(mut self, r: Region) -> Self {
+        self.region = r;
+        self
+    }
+
+    /// To view the available sizes in the relevant region, use:
+    /// `az vm list-sizes -l <region_name>`.
+    pub fn instance_type(mut self, inst_type: String) -> Self {
+        self.instance_type = inst_type;
+        self
+    }
+
+    /// Set the image.
+    ///
+    /// See `az vm image list` for valid options.
+    pub fn image(mut self, image: String) -> Self {
+        self.image = image;
+        self
+    }
+
+    pub fn username(mut self, username: String) -> Self {
+        self.username = username;
+        self
+    }
+
+    /// The provided callback, `setup`, is called once for every spawned instances of this type with a handle
+    /// to the target machine. Use [`Machine::ssh`] to issue
+    /// commands on the host in question.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use tsunami::providers::azure::Setup;
+    ///
+    /// let m = Setup::default()
+    ///     .setup(|ssh, log| {
+    ///         slog::info!(log, "running setup!");
+    ///         ssh.cmd("sudo apt update")?;
+    ///         Ok(())
+    ///     });
+    /// ```
+    pub fn setup(
+        mut self,
+        setup: impl Fn(&mut ssh::Session, &slog::Logger) -> Result<(), Error> + Send + Sync + 'static,
+    ) -> Self {
+        self.setup_fn = Some(Arc::new(setup));
+        self
+    }
+}
+
+/// Launcher type for the Microsoft Azure cloud.
+///
+/// This implementation relies on the [Azure
+/// CLI](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli?view=azure-cli-latest).
+///
+/// It also assumes you have previously run `az login` to authenticate.
+#[derive(Debug, Default)]
+pub struct Launcher {
```

Comment:
It'd probably be good to point people towards `TsunamiBuilder::spawn` here, since that's how they're likely to _use_ this type.